File: src/examples/C#/GroundWrapper/PathCache.cs
using System.Threading;

namespace Ground {

    public class PathCache {
        public PathCache(int capacity) {
            vertices = new PathVertex[capacity];
        }

        public PathVertex this[int vertexId] => vertices[vertexId];

        public int AddVertex(PathVertex vertex) {
            int idx = Interlocked.Increment(ref next);

            if (idx > vertices.Length)
                return -1;

            vertices[idx] = vertex;
            return idx;
        }

        public void Clear() {
            int overflow = next - vertices.Length;
            if (overflow > 0) {
                System.Console.WriteLine($"Overflow detected. Resizing to fit {overflow * 2} additional vertices.");
                vertices = new PathVertex[vertices.Length + overflow * 2];
            }

            next = 0;
        }

        PathVertex[] vertices;
        int next = 0;
    }

}

File: src/renderground/utility/pathcache.cpp
#include "utility/pathcache.h"

#include <cassert>
#include <iostream>

namespace ground
{

PathCache::PathCache(int initialSize)
: vertexCache(initialSize)
{
    next = 0;
    overflow = 0;
    Clear();
}

int PathCache::Add(const PathVertex& vertex) {
    auto idx = next++;

    if (idx >= vertexCache.size()) {
        ++overflow;
        return -1;
    }

    vertexCache[idx] = vertex;
    return idx;
}

void PathCache::Clear() {
    if (overflow > 0) {
        vertexCache.resize(vertexCache.size() + overflow * 2);

        // TODO should put this in some simple logging tool with verbosity levels.
        std::cout << "Info: A path cache overflow discarded a total of "
                  << overflow << " vertices. "
                  << "Allocating additional memory (twice that number). "
                  << "The total size is now: " << vertexCache.size() << std::endl;
    }

    next = 0;
    overflow = 0;
}

} // namespace ground
